Service factories

tomphp opened the pull request:
Resolves #43

tomphp commented:
@localheinz there's some questionable naming in this PR - feel free to be picky ;-)

localheinz commented on README.md:
```diff
@@ -184,6 +184,59 @@ Configurator::apply()->configFromArray($config)->to($container);
+#### Service Factories
+
+If you require some addition additional logic when creating a service, you can
+define a Service Factory. A service factory is simply and invokable class which
```
@tomphp 

Should this be

```diff
-define a Service Factory. A service factory is simply and invokable class which
+define a Service Factory. A service factory is simply an invokable class which
```

instead?

localheinz commented on README.md:
```diff
@@ -184,6 +184,59 @@ Configurator::apply()->configFromArray($config)->to($container);
+        'username' => 'example_user',
+        'password' => 'example_password',
+    ],
+
```
@tomphp 

Should we remove this extra empty line, or is it here on purpose?

tomphp replied:
I guess it was intentional but happy to remove it.

localheinz commented on src/League/ServiceServiceProvider.php:
```diff
@@ -42,6 +42,16 @@ public function register()
     private function registerService(ServiceDefinition $definition)
     {
+        if ($definition->isFactory()) {
+            $service = $this->getContainer()->add(
```
@tomphp 

After assigning `$service` a value here, it is no longer used, should we remove it?

localheinz commented on src/ServiceDefinition.php:
```diff
@@ -23,6 +24,11 @@
     /**
+     * @var bool
+     */
+    private $factory;
```
@tomphp 

Consistent with what we got, but should we consider prepending `is` to these boolean fields?

Wasn't quite sure at first whether this was supposed to be the name of the factory.

localheinz commented on tests/acceptance/SupportsServiceConfig.php:
```diff
@@ -287,4 +288,32 @@ public function testItUsesTheStringIffSetterMethodArgumentsAreClassNames()
+    {
+        $config = [
+            'class_name' => ExampleClassWithArgs::class,
+
```
@tomphp 

Should we remove this extra empty line, or is it here on purpose?

tomphp commented:
@localheinz I've addressed the issues but it seems a forced push doesn't work with the new GitHub code review stuff - will bare that in mind in the future.

localheinz commented:
@tomphp 

Yes, it's a bit weird that the comments are still there, noticed it, too!
